design/one-pager-rate-limiting.md

```diff
@@ -0,0 +1,151 @@
+managed resource).
+
+See [`controller.go`] and [`default_rate_limiters.go`].
+
+### Concurrent Reconciles
+
+Each controller may process at most one reconcile concurrently.
+
+## Crossplane Rate Limits
+
+The controller-runtime defaults are not suitable for Crossplane. Crossplane
+wants:
+
+* To wait more than 5ms before requeuing, but less than 16 minutes.
+* To reconcile several managed resources of a particular kind at once.
+* To rate limit _classes_ of managed resource (e.g. all AWS resources, or all
+  EC2 resources).
+
+Crossplane attempts to achieve this by deriving several rate limits from a
+single flag - `--max-reconcile-rate`. The default value for this flag is usually
+10 reconciles per second. The flag applies to an entire controller manager (e.g.
+Crossplane, or a provider).
+
+Note that provider maintainers must use the functions defined in [`default.go`]
+to ensure these rate limits are applied at the client, global, and controller
+levels.
+
+### API Server Request Rate
+
+An API server client that rate limits itself to `--max-reconcile-rate * 5` qps,
+and `--max-reconcile-rate * 10` burst. With a default `--max-reconcile-rate` of
+10 this is 50 qps bursting to 100 queries. This client is shared by all
+controllers that are part of the same controller manager (e.g. same provider).
+See [`default.go`].
+
+### Reconcile Rate
+
+Crossplane uses two layers of rate limiting.
+
+A global token bucket rate limiter limits all controllers within a provider to
+`--max-reconcile-rate` reconciles per second, bursting to
+`--max-reconcile-rate * 10`. With a default `--max-reconcile-rate` of 10 this is
+10 reconciles per second, bursting to 100.
+
+All reconciles are subject to the global rate limiter, even those triggered by a
+previous reconcile returning `reconcile.Result{RequeueAfter: t}`.
+
+An exponential backoff rate limiter limits how frequently a particular object
+may be reconciled, backing off from 1s to 60s. A reconcile triggered by a
+previous reconcile returning `reconcile.Result{RequeueAfter: t}` is not subject
+to this rate limiter.
+
+Due to limitations of controller-runtime (see [issue #857]) the global rate
+limiter is implemented as a middleware `Reconciler`. See [`reconciler.go`].
+
+Reconciles may be rate limited by both layers.
+
+Consider a reconcile that was requeued because it returned an error. First it's
+subject to the controller's exponential backoff reconciler, which adds the
+reconcile to the controller's work queue to be processed from 1 to 60 seconds in
+the future.
+
+When the reconcile is popped from the head of the work queue it's processed by
+the middleware `Reconciler`, subject to its token bucket reconciler. If there
+are sufficient tokens available in the bucket, the reconcile is passed to the
+wrapped (inner) `Reconciler` immediately. If there aren't sufficient tokens
+available, the reconcile is returned to the tail of the work queue by returning
```

Comment:
Wouldn't it be more fair if we put it head of the queue? Why do we change the ordering in the queue?
Maybe I misunderstood something but, as an analogy, I am waiting in a bank queue and when it is my turn I am told by the officer to go to the tail of the queue because he processed enough for that given hour 😅

Reply:
> Wouldn't it be more fair if we put it back to the head of the queue?

It would.

> Why do we change the ordering in the queue?

No reason, really. 🙂 It just seems to be how all the Kubernetes work queue rate limiting works, and the topic didn't come up until now.

The queueing is all using https://pkg.go.dev/k8s.io/client-go/util/workqueue. When we return `RequeueAfter` I 
believe that translates to an `AddAfter`, i.e. "add back to the end of the queue after duration D":

https://github.com/crossplane/crossplane-runtime/blob/v1.15.1/pkg/ratelimiter/reconciler.go#L51

We could potentially improve on this by using a `RateLimitingQueue` inside our `ratelimiter.Reconciler`. This way there would be two levels of workqueue:

* Outer queue handles per-object exponential backoff. Requests sent to back of outer queue.
* Inner queue handles global max-reconcile-rate. Requests sent to back of inner queue.

I think if we wanted to support requests being sent to the front of the queue after a certain duration we'd need to implement our own work queue.

There's talk of allowing users to supply their own work queue (not only their own rate limiter) in controller-runtime in https://github.com/kubernetes-sigs/controller-runtime/issues/857. If we want to optimize as best as possible, it might be worth pursuing that upstream.